src/lib.rs

```diff
@@ -3466,6 +3467,251 @@ fn digraph_complement(
+        py,
+        &graph.graph,
+        pos,
+        fixed,
+        k,
+        p,
+        adaptive_cooling,
+        niter,
+        tol,
+        weight_fn,
+        default_weight,
+        scale,
+        center,
+        seed,
+    )
+}
+
+/// Position nodes using Fruchterman-Reingold force-directed algorithm.
+///
+/// The algorithm simulates a force-directed representation of the network
+/// treating edges as springs holding nodes close, while treating nodes
+/// as repelling objects, sometimes called an anti-gravity force.
+/// Simulation continues until the positions are close to an equilibrium.
+//
+/// :param PyDiGraph graph: Graph to be used.
+/// :param dict (default=None) pos:
+///     Initial node positions as a dictionary with node ids as keys and
+///     values as a coordinate list. If None, then use random initial positions.
+/// :param set (default=None) fixed: Nodes to keep fixed at initial position.
+///     Error raised if fixed specified and pos not.
+/// :param float (default=None) k:
+///     Optimal distance between nodes. If None the distance is set to
+///     1/sqrt(n) where n is the number of nodes.  Increase this value
+///     to move nodes farther apart.
+/// :param int (default=2) p:
+///     Repulsive force exponent.
+/// :param bool (default=True) adaptive_cooling:
+///     Use an adaptive cooling scheme. If set to False,
+///     a linear cooling scheme is used.
+/// :param int (default=50) niter:
+///     Maximum number of iterations.
+/// :param float (default = 1e-6) tol:
+//      Threshold for relative error in node position changes.
+///     The iteration stops if the error is below this threshold.
+/// :param weight_fn: An optional weight function for an edge. It will accept
+///     a single argument, the edge's weight object and will return a float
+///     which will be used to represent the weight of the edge.
+/// :param float (default=1) default_weight: If ``weight_fn`` isn't specified
+///     this optional float value will be used for the weight/cost of each edge
+/// :param float or None (default=1) scale: Scale factor for positions.
+///     Not used unless fixed is None. If scale is None, no rescaling is performed.
+/// :param list center (default=None) – Coordinate pair around which to center
+///     the layout. Not used unless fixed is None.
+/// :param int seed: An optional seed to use for the random number generator
+///
+/// :returns: A dictionary of positions keyed by node id.
+/// :rtype: dict
+#[pyfunction]
+#[text_signature = "(graph, pos=None, fixed=None, k=None, p=2, adaptive_cooling=True,
+                     niter=50, tol=1e-6, weight_fn=None, default_weight=1, scale=1,
+                     center=None, seed=None, /)"]
+#[allow(clippy::too_many_arguments)]
+pub fn digraph_spring_layout(
+    py: Python,
+    graph: &digraph::PyDiGraph,
+    pos: Option<HashMap<usize, layout::Point>>,
+    fixed: Option<HashSet<usize>>,
+    k: Option<f64>,
+    p: Option<i32>,
```

Comment:
I realize these are straight from the paper, but I do wonder if we can think of more descriptive argument names in case someone hasn't read the paper. Not really a blocker, but something to consider.

Reply:
`k` is also used by networkx, so it might be better to leave it as it is. 
what about `repulsive_exponent` for `p`?

Reply:
Sure that works for me